tortoise/fixture_test.go

```diff
@@ -0,0 +1,791 @@
+	for i := 0; i < n; i++ {
+		s.runRandomTopo()
+	}
+}
+
+func (s *session) runRandomTopo() {
+	// TODO(dshulyak) compute and execute all topological paths
+	// or optionally a subset of them if all takes too long
+	rst := []action{}
+	empty := []action{}
+	forward := map[action]map[action]struct{}{}
+	reverse := map[action]map[action]struct{}{}
+	for _, act := range s.actions {
+		for _, dep := range act.deps() {
+			if reverse[act] == nil {
+				reverse[act] = map[action]struct{}{}
+			}
+			if forward[dep] == nil {
+				forward[dep] = map[action]struct{}{}
+			}
+			reverse[act][dep] = struct{}{}
+			forward[dep][act] = struct{}{}
+		}
+		if len(act.deps()) == 0 {
+			empty = append(empty, act)
+		}
+	}
+	rand.Shuffle(len(empty), func(i, j int) {
+		empty[i], empty[j] = empty[j], empty[i]
+	})
+	for len(empty) > 0 {
+		act := empty[0]
+		empty = empty[1:]
+		rst = append(rst, act)
+		for dep := range forward[act] {
+			delete(reverse[dep], act)
+			if len(reverse[dep]) == 0 {
+				delete(reverse, dep)
+				empty = append(empty, dep)
+			}
+		}
+		delete(forward, act)
+	}
+	require.Empty(s.tb, reverse, "all dependencies must be satifisied")
+	s.runActions(rst)
+}
+
+func (s *session) ensureConfig() {
+	if s.config == nil {
+		config := DefaultConfig()
+		s.config = &config
+	}
+}
+
+func (s *session) withEpochSize(val uint32) *session {
+	s.epochSize = int(val)
+	types.SetLayersPerEpoch(val)
+	return s
+}
+
+func (s *session) withHdist(val uint32) *session {
+	s.ensureConfig()
+	s.config.Hdist = val
+	return s
+}
+
+func (s *session) withZdist(val uint32) *session {
+	s.ensureConfig()
+	s.config.Zdist = val
+	return s
+}
+
+func (s *session) withWindow(val uint32) *session {
+	s.ensureConfig()
+	s.config.WindowSize = val
+	return s
+}
+
+func (s *session) withLayerSize(val uint32) *session {
+	s.ensureConfig()
+	s.layerSize = int(val)
+	s.config.LayerSize = val
+	return s
+}
+
+func (s *session) withDelay(val uint32) *session {
+	s.ensureConfig()
+	s.config.BadBeaconVoteDelayLayers = val
+	return s
+}
+
+func (s *session) tortoise() *Tortoise {
+	s.ensureConfig()
+	trt, err := New(WithLogger(logtest.New(s.tb)), WithConfig(*s.config))
+	require.NoError(s.tb, err)
+	return trt
+}
+
+func TestSanity(t *testing.T) {
+	const n = 2
+	var activeset []*atxAction
+	s := newSession(t)
+	for i := 0; i < n; i++ {
+		activeset = append(
+			activeset,
+			s.smesher(i).atx(1, new(aopt).height(100).weight(400)),
+		)
+	}
+	s.beacon(1, "a")
+	for i := 0; i < n; i++ {
+		s.smesher(i).atx(1).ballot(1, new(bopt).
+			beacon("a").
+			activeset(activeset...).
+			eligibilities(s.layerSize/n))
+	}
+	s.tally(1)
+	s.hareblock(1, "aa", 0)
+	for i := 0; i < n; i++ {
+		s.smesher(i).atx(1).ballot(2, new(bopt).
+			eligibilities(s.layerSize/n).
+			votes(new(evotes).support(1, "aa", 0)),
+		)
+	}
+	s.hareblock(2, "bb", 0)
+	s.tallyWait(2)
+	s.updates(t, new(results).
+		verified(0).
+		verified(1).block("aa", 0, valid|hare|data).
+		next(2).block("bb", 0, hare|data),
```

Comment:
is there a scenario where `hare` and `data` are different value?

Reply:
you mean in whole application? data can be required even if block is not confirmed by hare, hare might be true if there is a bug elsewhere and node didn't save the block but saved hare output